We can't connect to Snowflake via an SSH tunnel

### Describe the bug

Seems that we're expecting a port, which we don't provide on the snowflake connection page

### To Reproduce

1) set up Metabase and a bastion host
2) try to connect to a snowflake server via that bastion host
3) see the connection failing with the error below

### Expected behavior

We should be able to connect

### Logs

```
2025-03-26 22:50:04,442 ERROR api.database :: Cannot connect to Database
clojure.lang.ExceptionInfo: Assert failed: (integer? port) {:message "Assert failed: (integer? port)"}
	at metabase.driver.util$can_connect_with_details_QMARK_.invokeStatic(util.clj:190)
	at metabase.driver.util$can_connect_with_details_QMARK_.doInvoke(util.clj:161)
	at clojure.lang.RestFn.invoke(RestFn.java:445)
	at metabase.api.database$test_database_connection.invokeStatic(database.clj:731)
	at metabase.api.database$test_database_connection.doInvoke(database.clj:721)
	at clojure.lang.RestFn.invoke(RestFn.java:428)
	at metabase.api.database$test_connection_details106669__106670.invokeStatic(database.clj:785)
	at metabase.api.database$test_connection_details106669__106670.invoke(database.clj:765)
	at metabase.api.database$fn__106677$fn__106683.invoke(database.clj:809)
	at metabase.api.macros$_core_fn92702__92703$core_fn__92704.invoke(macros.clj:405)
	at metabase.api.macros$endpoint_handler_STAR_92746__92748$handler__92752.invoke(macros.clj:601)
	at metabase.api.macros$build_ns_handler92779__92780$ns_handler_STAR___92781.invoke(macros.clj:665)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.server.middleware.auth$enforce_authentication$fn__55706.invoke(auth.clj:18)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__103223.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160$f__103161$respond_SINGLEQUOTE___103162.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__103223.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160.invoke(core.clj:200)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.server.routes$fn__112217$fn__112218.invoke(routes.clj:70)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160.invoke(core.clj:200)
	at compojure.core$make_context$handler__103188.invoke(core.clj:290)
	at compojure.core$make_context$fn__103192.invoke(core.clj:300)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160$f__103161$respond_SINGLEQUOTE___103162.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103141.invoke(core.clj:153)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160$f__103161$respond_SINGLEQUOTE___103162.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103141.invoke(core.clj:153)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160$f__103161$respond_SINGLEQUOTE___103162.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103141.invoke(core.clj:153)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160$f__103161$respond_SINGLEQUOTE___103162.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103141.invoke(core.clj:153)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160$f__103161$respond_SINGLEQUOTE___103162.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__103223.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__103223.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at compojure.core$routes$fn__103160$f__103161.invoke(core.clj:198)
	at compojure.core$routes$fn__103160.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__56046.invoke(exceptions.clj:111)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__56043.invoke(exceptions.clj:100)
	at metabase.server.middleware.log$log_api_call$fn__119181$fn__119182$fn__119183.invoke(log.clj:232)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invokeStatic(diagnostic.clj:17)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invoke(diagnostic.clj:12)
	at metabase.server.middleware.log$log_api_call$fn__119181$fn__119182.invoke(log.clj:223)
	at toucan2.execute$do_with_call_counts.invokeStatic(execute.clj:112)
	at toucan2.execute$do_with_call_counts.invoke(execute.clj:103)
	at metabase.server.middleware.log$log_api_call$fn__119181.invoke(log.clj:222)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__127549.invoke(browser_cookie.clj:40)
	at metabase.server.middleware.security$add_security_headers$fn__56002.invoke(security.clj:273)
	at metabase.server.middleware.json$wrap_json_body$fn__48809.invoke(json.clj:128)
	at metabase.server.middleware.offset_paging$handle_paging$fn__126347.invoke(offset_paging.clj:48)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__48797.invoke(json.clj:83)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__127816.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__127835.invoke(params.clj:77)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__118599.invoke(misc.clj:59)
	at metabase.server.middleware.session$reset_session_timeout$fn__121140.invoke(session.clj:270)
	at metabase.server.middleware.session$bind_current_user$fn__121133$fn__121134.invoke(session.clj:245)
	at metabase.request.session$do_with_current_user.invokeStatic(session.clj:47)
	at metabase.request.session$do_with_current_user.invoke(session.clj:30)
	at metabase.server.middleware.session$bind_current_user$fn__121133.invoke(session.clj:244)
	at metabase.server.middleware.session$wrap_current_user_info$fn__121125.invoke(session.clj:204)
	at metabase.analytics.sdk$embedding_mw$embedding_mw_fn__79544.invoke(sdk.clj:64)
	at metabase.server.middleware.session$wrap_session_id$fn__121097.invoke(session.clj:82)
	at metabase.server.middleware.auth$wrap_static_api_key$fn__55714.invoke(auth.clj:33)
	at ring.middleware.cookies$wrap_cookies$fn__127736.invoke(cookies.clj:200)
	at metabase.server.middleware.misc$add_content_type$fn__118581.invoke(misc.clj:28)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__118607.invoke(misc.clj:75)
	at ring.middleware.gzip$wrap_gzip$fn__127778.invoke(gzip.clj:86)
	at metabase.server.middleware.request_id$wrap_request_id$fn__116502.invoke(request_id.clj:10)
	at metabase.server.middleware.misc$bind_request$fn__118610$fn__118611.invoke(misc.clj:87)
	at metabase.request.current$do_with_current_request.invokeStatic(current.clj:19)
	at metabase.request.current$do_with_current_request.invoke(current.clj:15)
	at metabase.server.middleware.misc$bind_request$fn__118610.invoke(misc.clj:86)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__127565.invoke(ssl.clj:51)
	at clojure.lang.Var.invoke(Var.java:395)
	at metabase.server.instance$async_proxy_handler$fn__56386.invoke(instance.clj:76)
	at metabase.server.instance.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel$RequestDispatchable.dispatch(HttpChannel.java:1598)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:753)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:501)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:287)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:421)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:390)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:277)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:199)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:411)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.AssertionError: Assert failed: (integer? port)
	at metabase.util.ssh$start_ssh_tunnel_BANG_.invokeStatic(ssh.clj:73)
	at metabase.util.ssh$start_ssh_tunnel_BANG_.invoke(ssh.clj:73)
	at metabase.util.ssh$include_ssh_tunnel_BANG_.invokeStatic(ssh.clj:115)
	at metabase.util.ssh$include_ssh_tunnel_BANG_.invoke(ssh.clj:109)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:158)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:154)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invokeStatic(connection.clj:375)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invoke(connection.clj:371)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:403)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:399)
	at metabase.driver.sql_jdbc$fn__128842.invokeStatic(sql_jdbc.clj:50)
	at metabase.driver.sql_jdbc$fn__128842.invoke(sql_jdbc.clj:48)
	at metabase.driver.snowflake$fn__120319.invokeStatic(snowflake.clj:728)
	at metabase.driver.snowflake$fn__120319.invoke(snowflake.clj:726)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__58127.invoke(util.clj:173)
	at clojure.core$binding_conveyor_fn$fn__5842.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
```

### Information about your Metabase installation

```JSON
v53.5 but it's been like that since... forever?
```

### Severity

P1

### Additional context

NA

--- a/src/metabase/driver/sql_jdbc/connection.clj
+++ b/src/metabase/driver/sql_jdbc/connection.clj
@@ -371,7 +371,10 @@ For setting the maximum, see [MB_APPLICATION_DB_MAX_CONNECTION_POOL_SIZE](#mb_ap
 (defn do-with-connection-spec-for-testing-connection
   "Impl for [[with-connection-spec-for-testing-connection]]."
   [driver details f]
-  (let [details (update details :port #(or % (default-ssh-tunnel-target-port driver)))]
+  (let [details (driver/incorporate-ssh-tunnel-details
+                  ;; If the tunnel is disabled this returned unchanged
+                 driver
+                 (update details :port #(or % (default-ssh-tunnel-target-port driver))))]
     (ssh/with-ssh-tunnel [details-with-tunnel details]
       (let [details-with-auth (driver.u/fetch-and-incorporate-auth-provider-details
                                driver

--- a/test/metabase/driver/sql_jdbc/connection_test.clj
+++ b/test/metabase/driver/sql_jdbc/connection_test.clj
@@ -372,6 +372,8 @@
                                      :tunnel-port ssh-test/ssh-mock-server-with-password-port
                                      :tunnel-user ssh-test/ssh-username
                                      :tunnel-pass ssh-test/ssh-password)]
+
+        (is (true? (driver.u/can-connect-with-details? (tx/driver) tunnel-db-details)))
         (mt/with-temp [:model/Database tunneled-db {:engine (tx/driver), :details tunnel-db-details}]
           (mt/with-db tunneled-db
             (sync/sync-database! (mt/db))

